paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java

```diff
@@ -18,168 +18,110 @@
-            return requestedMode;
-        } else if (table instanceof ChangelogValueCountFileStoreTable) {
-            // no primary key, sink all changelogs
-            return requestedMode;
-        } else if (table instanceof ChangelogWithKeyFileStoreTable) {
+    public RowLevelUpdateInfo applyRowLevelUpdate(
+            List<Column> updatedColumns, @Nullable RowLevelModificationScanContext context) {
+        // Since only UPDATE_AFTER type messages can be received at present,
+        // AppendOnlyFileStoreTable and ChangelogValueCountFileStoreTable without primary keys
+        // cannot correctly handle old data, so they are marked as unsupported. Similarly, it is not
+        // allowed to update the primary key column when updating the column of
+        // ChangelogWithKeyFileStoreTable, because the old data cannot be handled correctly.
+        if (table instanceof ChangelogWithKeyFileStoreTable) {
             Options options = Options.fromMap(table.options());
-            if (options.get(CHANGELOG_PRODUCER) == ChangelogProducer.INPUT) {
-                return requestedMode;
-            }
-
-            if (options.get(MERGE_ENGINE) == MergeEngine.AGGREGATE) {
-                return requestedMode;
-            }
-
-            if (options.get(LOG_CHANGELOG_MODE) == LogChangelogMode.ALL) {
-                return requestedMode;
-            }
-
-            // with primary key, default sink upsert
-            ChangelogMode.Builder builder = ChangelogMode.newBuilder();
-            for (RowKind kind : requestedMode.getContainedKinds()) {
-                if (kind != RowKind.UPDATE_BEFORE) {
-                    builder.addContainedKind(kind);
+            Set<String> primaryKeys = new HashSet<>(table.primaryKeys());
+            updatedColumns.forEach(
+                    column -> {
+                        if (primaryKeys.contains(column.getName())) {
+                            throw new UnsupportedOperationException(
+                                    "Updates to primary keys are not supported.");
+                        }
+                    });
+
+            if (options.get(MERGE_ENGINE) == MergeEngine.DEDUPLICATE) {
+                return new RowLevelUpdateInfo() {};
+            } else if (options.get(MERGE_ENGINE) == MergeEngine.PARTIAL_UPDATE) {
+                this.updatedColumns = new LinkedHashMap<>();
+                for (int i = 0; i < columns.size(); i++) {
+                    Column column = columns.get(i);
+                    if (primaryKeys.contains(column.getName()) || updatedColumns.contains(column)) {
+                        this.updatedColumns.put(
+                                i,
+                                RowData.createFieldGetter(
+                                        column.getDataType().getLogicalType(), i));
+                    }
                 }
+                // Even with partial-update we still need all columns. Because the topology
```

Comment:
Can you create a Flink JIRA issue?

Reply:
Created a related Flink issue.  https://issues.apache.org/jira/browse/FLINK-32001